_posts/2022-09-19-blog-wpe-networking-overview.md

```diff
@@ -0,0 +1,49 @@
+
+### WebKit
+
+Before we get into the libraries used for WPE lets discuss WebKit itself. Despite abstracting out a lot of the protocol handling WebKit itself still needs to understand a lot of fundamentals of HTTP.
+
+WebCore (discussed in [WPE Overview](/blog/02-overview-of-wpe.html)) understands HTTP requests, headers, and cookies are as they are required to implement many higher level features. What it does not do is the network operations, most parsing, or on-disk storage for them. In the codebase these are represented by `ResourceRequest` and `ResourceResponse` objects mapping to general HTTP functionality.
+
+#### NetworkProcess
+
+A core part of modern web engine security is the multi-process model. Each website runs in its own isolated process that does not have network access to defend from exploits. In order to allow for network access they must talk over IPC to a dedicated NetworkProcess, one per browser instance typically. The NetworkProcess receives a `ResourceRequest`, creates a `NetworkDataTask` with it to download the data, and responds with a `ResourceResponse` to the WebProcess.
+
+This is where the bulk of interaction with system networking libraries happens. In the case of the WPE port it uses libsoup and GLib. WPE provides platform specific implementations of these objects: `ResourceRequestSoup` and `NetworkDataTaskSoup`.
+
+### libsoup
+
+The [libsoup](https://libsoup.org) library was originally created for the GNOME project for its email client and has since grown to be a very featureful HTTP implementation now maintained by Igalia.
+
+At a high level the main task that libsoup does is manage connections and queued requests to websites and then efficiently stream the responses back to WPE. However properly implementing HTTP is a fairly large task and this is a non-exhaustive list of features it implements: HTTP/1.1, HTTP/2, WebSockets, cookies (including on-disk database), decompression, multiple authentication standards, HSTS, and HTTP proxies.
+
+### GLib
+
+libsoup on its own is really focused on the HTTP layer and uses the [GLib](https://gitlab.gnome.org/GNOME/glib) library to implement many of its networking features in a portable way. This is where TCP, DNS, and TLS are handled. It is also directly used by WebKit for URI parsing and DNS pre-caching.
+
+Using GLib also helps standardize behavior across a modern Linux system. It allows configuration of a global proxy resolver that WebKit along with other applications will use.
+
+#### TLS
+
+TLS brings its own complexities and is actually abstracted out into another project called [GLib-Networking](https://gitlab.gnome.org/GNOME/glib-networking). This project provides multiple implementations of TLS that can be chosen at runtime including OpenSSL and gnutls. The benefits to this is clients can choose the implementation they prefer be it for licensing, certification, or technical reasons.
```

Comment:
'prefer be it' needs broken up I think... just a comma after prefer maybe?  @meyerweb probably has an answer here

Reply:
I reworded this a bit in my edit pass.